substratevm/src/com.oracle.objectfile/src/com/oracle/objectfile/elf/dwarf/DwarfLocSectionImpl.java

```diff
@@ -0,0 +1,718 @@
+                if (kind == JavaKind.Int) {
+                    pos += putByte((byte) constant.asInt(), scratch, 0);
+                } else {
+                    pos += putByte((byte) (constant.asBoolean() ? 1 : 0), scratch, 0);
+                }
+            } else if (byteCount == 2) {
+                pos += putShort((short) constant.asInt(), scratch, 0);
+            } else if (byteCount == 4) {
+                int i = (kind == JavaKind.Int ? constant.asInt() : Float.floatToRawIntBits(constant.asFloat()));
+                pos += putInt(i, scratch, 0);
+            } else {
+                long l = (kind == JavaKind.Long ? constant.asLong() : Double.doubleToRawLongBits(constant.asDouble()));
+                pos += putLong(l, scratch, 0);
+            }
+        } else {
+            pos = putByte(op, buffer, pos);
+            pos = putULEB(byteCount, buffer, pos);
+            if (constant.isNull()) {
+                pos = writeAttrData8(0, buffer, pos);
+            } else if (byteCount == 1) {
+                if (kind == JavaKind.Int) {
+                    pos = putByte((byte) constant.asInt(), buffer, pos);
+                } else {
+                    pos = putByte((byte) (constant.asBoolean() ? 1 : 0), buffer, pos);
+                }
+            } else if (byteCount == 2) {
+                pos = putShort((short) constant.asInt(), buffer, pos);
+            } else if (byteCount == 4) {
+                int i = (kind == JavaKind.Int ? constant.asInt() : Float.floatToRawIntBits(constant.asFloat()));
+                pos = putInt(i, buffer, pos);
+            } else {
+                long l = (kind == JavaKind.Long ? constant.asLong() : Double.doubleToRawLongBits(constant.asDouble()));
+                pos = putLong(l, buffer, pos);
+            }
+            verboseLog(context, "  [0x%08x]     CONSTANT %s", pos, constant.toValueString());
+        }
+        return pos;
+    }
+
+    // auxiliary class used to collect per-range locations for a given local
+    // merging adjacent ranges with the same location
+    static class LocalValueExtent {
+        long lo;
+        long hi;
+        DebugLocalValueInfo value;
+
+        LocalValueExtent(long lo, long hi, DebugLocalValueInfo value) {
+            this.lo = lo;
+            this.hi = hi;
+            this.value = value;
+        }
+
+        @SuppressWarnings("unused")
+        boolean shouldMerge(int otherLo, int otherHi, DebugLocalValueInfo otherValue) {
+            // ranges need to be contiguous to merge
+            if (hi != otherLo) {
+                return false;
+            }
+            // values need to be for the same line
+            if (value.line() != otherValue.line()) {
+                return false;
+            }
+            // location kinds must match
+            if (value.localKind() != otherValue.localKind()) {
+                return false;
+            }
+            // locations must match
+            switch (value.localKind()) {
```

Comment:
Shouldn't all those checks on DebugLocalValueInfos be implemented by providing an `public boolean equals(Object obj)` on `NativeImageDebugLocalValueInfo`. Also there seems to be some redundancy between here and `com.oracle.svm.hosted.image.NativeImageDebugInfoProvider.NativeImageDebugLocalValueInfo#sameAs`

Reply:
Yes, I think so. However, note that the above check is happening in the generator code. So, as well as implementing `boolean equals(Object obj)` in class `NativeImageDebugLocalValueInfo` we would also need to expose it as a method of interface `DebugLocalValueInfo`.

Reply:
> we would also need to expose it as a method of interface DebugLocalValueInfo

Not sure why. Every Java object has `boolean equals(Object obj)` implicitly as part of its API. It's just that the default implementation is based of ref-equality instead of contents-equality. If you change this it an `equals` override in NativeImageDebugLocalValueInfo please also add `hashcode` for completeness (they should always come in pairs).

Reply:
Doh! ... :-)

Reply:
Fixed.